includes/data-port/class-sensei-data-port-job.php

```diff
@@ -0,0 +1,332 @@
+<?php
+/**
+ * File containing the Sensei_Data_Port_Job class.
+ *
+ * @package sensei
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Superclass of data import/export jobs. It provides basic functionality like logging, maintaining state, cleanup and
+ * running data port tasks which are registered by subclasses.
+ */
+abstract class Sensei_Data_Port_Job implements Sensei_Background_Job_Interface {
+	const OPTION_PREFIX = 'sensei-tools-';
+
+	/**
+	 * An array which holds the results of the data port job and populated in subclasses.
+	 *
+	 * @var array
+	 */
+	protected $results;
+
+	/**
+	 * An array which contains the state for each task.
+	 *
+	 * @var array
+	 */
+	protected $task_state;
+
+	/**
+	 * Unique id for the job.
+	 *
+	 * @var string
+	 */
+	private $job_id;
+
+	/**
+	 * An array containing log entries (e.g. errors). It has the following format:
+	 * {
+	 *
+	 *     @type array $type {
+	 *         @type string $title The post title of the entity this entry applies to.
+	 *         @type string $id    The id of the entity this entry applies to.
+	 *         @type string $msg   The log message.
+	 *     }
+	 * }
+	 *
+	 * @var array
+	 */
+	private $logs;
+
+	/**
+	 * True if the job is completed.
+	 *
+	 * @var bool
+	 */
+	private $is_completed;
+
+	/**
+	 * True if the internal state has changed and needs to be stored.
+	 *
+	 * @var bool
+	 */
+	private $has_changed;
+
+	/**
+	 * Estimate of completion percentage.
+	 *
+	 * @var float
+	 */
+	private $percentage;
+
+	/**
+	 * The tasks that consist this data port job.
+	 *
+	 * @var Sensei_Data_Port_Task_Interface[]
+	 */
+	private $tasks;
+
+	/**
+	 * Sensei_Data_Port_Job constructor.
+	 *
+	 * @param string $job_id   Unique job id.
+	 * @param string $json     A json string to restore internal state from.
+	 */
+	protected function __construct( $job_id, $json = '' ) {
+		$this->job_id      = $job_id;
+		$this->has_changed = false;
+
+		if ( '' !== $json ) {
+			$this->restore_from_json( $json );
+		} else {
+			$this->logs         = [];
+			$this->results      = [];
+			$this->is_completed = false;
+			$this->has_changed  = true;
+			$this->task_state   = [];
+			$this->percentage   = 0;
+		}
+
+		$this->tasks = $this->get_tasks();
```

Comment:
Are we leaving the responsibility for restoring a task's state to `get_tasks`? I imagine there could be some shared state, as well, but task objects will probably be aware of this job instance they can do things with it.

It is a bit strange to set a private/protected attribute with a getter named with the attribute it is setting. Usually getters expose private/protected properties. I'm not sure what another name could be though without overly complicating it so maybe it is fine.

Reply:
> Are we leaving the responsibility for restoring a task's state to get_tasks?

I thought that this approach is the most agile at this point and we should probably leave it for the time that we started implementing tasks.

> It is a bit strange to set a private/protected attribute with a getter named with the attribute it is setting.

Agreed. I am working on another PR now which will introduce the manager class for the data port jobs as I thought that a good way to spot mistakes in the class is to start using it :smile:. In my new PR I have removed the `$tasks` property entirely and I moved it in the subclasses.